woocommerce/google-listings-and-ads
# Product Block Editor: Add the fundamental support and map `Text` and `Integer` attribute inputs to currently available generic blocks

### Changes proposed in this Pull Request:

To be compatible with Product Block Editor, this PR adds fundamental support:

- Extract the processing of getting the visible and hidden product types in `AttributesForm` as a method.
- Convert the `init_input` method in `AttributesForm`  to a static method.
- Add methods to `Admin\Input\Input` for getting block config.
- Add a new class `AttributesBlock` to register blocks to Product Block Editor for product attributes.

Also, it turns a few attribute inputs to support Product Block Editor first:

- Map `Text` and `Integer` inputs to currently available generic blocks.
- Add the block attribute 'min' to the **multipack** input.

Last, it adds some PHP tests and removes a few unused PHP classes or method:

- Remove the unused method `is_custom_value` from `SelectWithTextInput`.
- Add PHP unit tests for
   - the `Admin\Input\Input` class and its inheritance classes.
   - the attribute input classes under `Admin\Product\Attributes\Input`.
   - the `AttributesForm` class.
   - the `AttributesBlock` class.

### Screenshots:

#### 📷 Simple product

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/ff0c2831-df8b-4331-902a-719ac0999d59)

#### 📷 Variation product

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/5b5ea27e-5698-4e67-8410-1ccb70d5615a)


#### 📷 The `min` attribute of  the multipack input

![image](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/0c08606a-6519-4d39-9127-9405e3535813)

### Detailed test instructions:

#### 📌 <a href="#prepare-test-environment" id="prepare-test-environment">Prepare test environment</a>

1. Set up [WooCommerce 8.3.0](https://github.com/woocommerce/woocommerce/releases/tag/8.3.0)
2. Install [WooCommerce Beta Tester](https://woo.com/products/woocommerce-beta-tester/)
3. Go to WooCommerce > Settings > Advanced > Features. Enable the **New product editor**
   ![1](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/089066c0-2816-4ff6-a049-22d57fdc1ce6)
4. Go to Tools > WCA Test Helper > Features. Make sure the **product-block-editor** and **product-variation-management** options are enabled.
   ![2023-11-22 14 27 50](https://github.com/woocommerce/google-listings-and-ads/assets/17420811/8d57b85f-7c28-4c0c-aae5-748e9cd7abe8)
5. Deactivate WooCommerce Beta Tester as it's incompatible with GLA for now. See #2147

#### 📌 Test the simple product type

1. Go to edit a simple product, e.g. the **Beanie with Logo** imported from WC's sample products
2. Switch to the **Organization** tab and find the **Attributes** section.
3. Check if there is the **Google Listings & Ads** section under the **Attributes** section.
4. Within the **Google Listings & Ads** section, operate the text and number fields and change their values.
5. Click on the **Update** button at the top-right of the page.
6. Refresh the page or switch back to the classical product editor to see if the changed values are saved.

#### 📌 Test the variable and variation product type

1. Go to edit a variable product, e.g. the **Hoodie** imported from WC's sample products
2. Switch to the **Organization** tab and find the **Attributes** section.
3. Check if there is the **Google Listings & Ads** section under the **Attributes** section.
4. Within the **Google Listings & Ads** section, there should be only a brand field for now. Operate the brand field and change its value.
5. Click on the **Update** button at the top-right of the page.
6. Refresh the page or switch back to the classical product editor to see if the changed value is saved.
7. Back to edit the same variable product with Product Block Editor.
8. Switch to the **Variations** tab and find the **Variations** section.
9. Click on one of the Edit links within the **Variations** section to go to edit a variation product.
10. With the **General** tab, find the **Image** section.
11. Check if there is the **Google Listings & Ads** section under the **Image** section.
12. Repeat the similar test to verify the fields for this variation product.

#### 📌 Test the related filters

1. Continue to use the above variation product editing page.
2. Filters the list of applicable product types for each attribute - `"woocommerce_gla_attribute_applicable_product_types_{$attribute_id}"`
   1. Confirm there is no **Brand** field.
   2. Add a filter to the **google-listings-and-ads.php** file:
      ```php
      add_filter(
      	"woocommerce_gla_attribute_applicable_product_types_brand",
      	function ( array $applicable_types ) {
      		$applicable_types[] = 'variation';
      
      		return $applicable_types;
      	}
      );
      ```
   3. Refresh the page.
   4. Check if the **Brand** field appears.
3. Filters the list of product types to hide for each attribute - `"woocommerce_gla_attribute_hidden_product_types_{$attribute_id}"`
   1. Confirm there is the **Size** field.
   2. Add a filter to the **google-listings-and-ads.php** file:
      ```php
      add_filter(
      	"woocommerce_gla_attribute_hidden_product_types_size",
      	function ( array $applicable_types ) {
      		$applicable_types[] = 'variation';
      
      		return $applicable_types;
      	}
      );
      
      ```
   3. Refresh the page.
   4. Check if the **Size** field disappears.

### Additional details:

The generic number block (`'woocommerce/product-number-field'`) doesn't support verifying/limiting for entering an integer. If it's not supported afterward, then we may have to convert it to a custom block.

### Changelog entry
